src/api/apiService.js

```diff
@@ -2,53 +2,73 @@ import { sessionService } from 'redux-react-native-session';
 import humps from 'humps';
 import Config from 'react-native-config';
 
-const saveSessionHeaders = (headers) => {
-  if (headers.get('access-token')) {
-    const sessionHeaders = {
-      token: headers.get('access-token'),
-      uid: headers.get('uid'),
-      client: headers.get('client')
-    };
-    sessionService.saveSession(sessionHeaders);
-  }
+import handleErrors from './utils/handleErrors';
+import getResponseBody from './utils/getResponseBody';
+
+const ACCESS_TOKEN = 'access-token';
+const APPLICATION_JSON = 'application/json';
+const CONTENT_TYPE = 'Content-Type';
+
+const HTTP_VERB = {
+  GET: 'get',
+  POST: 'post',
+  DELETE: 'delete',
+  PUT: 'put',
+  PATCH: 'patch',
 };
 
-const handleErrors = response =>
-  new Promise((resolve, reject) => {
-    if (!response) {
-      reject({ message: 'No response returned from fetch' });
-      return;
-    }
+class Api {
+  static get(uri, apiUrl = Config.API_URL) {
+    const requestData = this.buildRequestData(HTTP_VERB.GET);
+    return this.loadHeadersAndPerformRequest(uri, apiUrl, requestData);
+  }
 
-    if (response.ok) {
-      saveSessionHeaders(response.headers);
-      resolve(response);
-      return;
-    }
+  static post(uri, data, apiUrl = Config.API_URL) {
+    const requestData = this.buildRequestData(HTTP_VERB.POST, data);
+    return this.loadHeadersAndPerformRequest(uri, apiUrl, requestData);
+  }
 
-    sessionService.loadSession()
-      .then(() => {
-        if (response.status === 401) {
-          sessionService.deleteSession();
-        }
-      });
+  static delete(uri, data, apiUrl = Config.API_URL) {
+    const requestData = this.buildRequestData(HTTP_VERB.DELETE, data);
+    return this.loadHeadersAndPerformRequest(uri, apiUrl, requestData);
+  }
 
-    response.json()
-      .then((json) => {
-        const error = json || { message: response.statusText };
-        reject(error);
-      }).catch(() => reject({ message: 'Response not JSON' }));
-  });
+  static put(uri, data, apiUrl = Config.API_URL) {
+    const requestData = this.buildRequestData(HTTP_VERB.PUT, data);
+    return this.loadHeadersAndPerformRequest(uri, apiUrl, requestData);
+  }
 
-const getResponseBody = (response) => {
-  const bodyIsEmpty = response.status === 204;
-  if (bodyIsEmpty) {
-    return Promise.resolve();
+  static patch(uri, data, apiUrl = Config.API_URL) {
+    const requestData = this.buildRequestData(HTTP_VERB.PATCH, data);
+    return this.loadHeadersAndPerformRequest(uri, apiUrl, requestData);
+  }
+
+  static buildRequestData(httpVerb, data = null) {
+    const requestData = {
+      method: httpVerb,
+      headers: {
+        accept: APPLICATION_JSON,
+        [CONTENT_TYPE]: APPLICATION_JSON,
+      }
+    };
+
+    if (data) {
+      const decamelizedData = humps.decamelizeKeys(data);
+      requestData.body = JSON.stringify(decamelizedData);
+    }
+
+    return requestData;
+  }
+
+  static loadHeadersAndPerformRequest(uri, apiUrl, data) {
```

Comment:
Use `async` / `await` and don't modify the input params.

```suggestion
  static async loadHeadersAndPerformRequest(uri, apiUrl, data) {
    const requestData = { ...data }
    try {
      const headers = await Api.getTokenHeader();
      requestData.headers = { ...requestData.headers, ...headers };
    } finally {
      return Api.performRequest(uri, apiUrl, requestData)
    }
  }
```

Reply:
I was using finally but i received an error from the lintern. Turns out that finally could led to some unexpected results, so i went with just a `catch`

https://eslint.org/docs/rules/no-unsafe-finally